node/src/client.rs

```diff
@@ -0,0 +1,396 @@
+    traits::{BlakeTwo256, Block as BlockT},
+    Justifications,
+};
+use sp_storage::{ChildInfo, PrefixedStorageKey, StorageData, StorageKey};
+use std::sync::Arc;
+
+/// A set of APIs that MainRT-like runtimes must implement.
+pub trait RuntimeApiCollection:
+    sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
+    + sp_api::ApiExt<Block>
+    + sp_consensus_babe::BabeApi<Block>
+    + pallet_grandpa::fg_primitives::GrandpaApi<Block>
+    + sp_block_builder::BlockBuilder<Block>
+    + frame_system_rpc_runtime_api::AccountNonceApi<Block, AccountId, Index>
+    + pallet_transaction_payment_rpc_runtime_api::TransactionPaymentApi<Block, Balance>
+    + sp_api::Metadata<Block>
+    + sp_offchain::OffchainWorkerApi<Block>
+    + sp_session::SessionKeys<Block>
+    + sp_authority_discovery::AuthorityDiscoveryApi<Block>
+where
+    <Self as sp_api::ApiExt<Block>>::StateBackend: sp_api::StateBackend<BlakeTwo256>,
+{
+}
+
+impl<Api> RuntimeApiCollection for Api
+where
+    Api: sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>
+        + sp_api::ApiExt<Block>
+        + sp_consensus_babe::BabeApi<Block>
+        + pallet_grandpa::fg_primitives::GrandpaApi<Block>
+        + sp_block_builder::BlockBuilder<Block>
+        + frame_system_rpc_runtime_api::AccountNonceApi<Block, AccountId, Index>
+        + pallet_transaction_payment_rpc_runtime_api::TransactionPaymentApi<Block, Balance>
+        + sp_api::Metadata<Block>
+        + sp_offchain::OffchainWorkerApi<Block>
+        + sp_session::SessionKeys<Block>
+        + sp_authority_discovery::AuthorityDiscoveryApi<Block>,
+    <Self as sp_api::ApiExt<Block>>::StateBackend: sp_api::StateBackend<BlakeTwo256>,
+{
+}
+
+/// Trait that abstracts over all available client implementations.
+///
+/// For a concrete type there exists [`Client`].
+pub trait AbstractClient<Block, Backend>:
+    BlockchainEvents<Block>
+    + Sized
+    + Send
+    + Sync
+    + ProvideRuntimeApi<Block>
+    + HeaderBackend<Block>
+    + CallApiAt<Block, StateBackend = Backend::State>
+where
+    Block: BlockT,
+    Backend: BackendT<Block>,
+    Backend::State: sp_api::StateBackend<BlakeTwo256>,
+    Self::Api: RuntimeApiCollection<StateBackend = Backend::State>,
+{
+}
+
+impl<Block, Backend, Client> AbstractClient<Block, Backend> for Client
+where
+    Block: BlockT,
+    Backend: BackendT<Block>,
+    Backend::State: sp_api::StateBackend<BlakeTwo256>,
+    Client: BlockchainEvents<Block>
+        + ProvideRuntimeApi<Block>
+        + HeaderBackend<Block>
+        + Sized
+        + Send
+        + Sync
+        + CallApiAt<Block, StateBackend = Backend::State>,
+    Client::Api: RuntimeApiCollection<StateBackend = Backend::State>,
+{
+}
+
+/// Execute something with the client instance.
+///
+/// As there exist multiple chains inside MainRT, like MainRT itself and testing runtime,
+/// there can exist different kinds of client types. As these client types differ in the generics
```

Comment:
Could we instead use the same client with all these different runtimes? Any api that the client should implement to match with a runtime, we can make sure the single client would implement all of them. It would then reduce the complexity of creating an matching two different clients for two different runtimes here.